### PR TITLE
Refactor atomic test cases to reduce code duplication

### PR DESCRIPTION
**Pull request by testhound**
1. Added new functions that encapsulate common test code behavior for binary operations
2. Rewrote test cases to utilize new functions

### CODE REVIEW

**Review by gmarkall — APPROVED** [review]
This looks good to me and tests out fine locally with the hardware target and the simulator.

It does highlight some weirdness about the tests - I'm not really sure that having variants of the tests with casts adds anything - perhaps they might be removed in future... @sklam do you know why e.g. there are two variants, `atomic_add2` and `atomic_add3`, where the only difference between them is the casting of the value assigned to `bin`?

**Comment by gmarkall** [comment]
@stuartarchibald Could this have a buildfarm run please?

**Comment by stuartarchibald** [comment]
Buildfarm ID `numba_smoketest_cuda_104`.

**Comment by stuartarchibald** [comment]
> It does highlight some weirdness about the tests - I'm not really sure that having variants of the tests with casts adds anything - perhaps they might be removed in future... @sklam do you know why e.g. there are two variants, `atomic_add2` and `atomic_add3`, where the only difference between them is the casting of the value assigned to `bin`?

Is this perhaps to exercise support for Tuple and a UniTuple as an index?

**Comment by stuartarchibald** [comment]
Buildfarm passed.

File: numba/cuda/tests/cudapy/test_atomics.py [hunk 1]
```diff
@@ -2,7 +2,7 @@
 import numpy as np
 from textwrap import dedent
 
-from numba import cuda, uint32, uint64, float32, float64
+from numba import cuda, uint32, uint64, float32, float64, void
```
Comment (stuartarchibald) [inline numba/cuda/tests/cudapy/test_atomics.py:5]: unused import

File: numba/cuda/tests/cudapy/test_atomics.py [hunk 2]
```diff
@@ -20,229 +20,154 @@ def skip_unless_cc_32(fn):
...
+    return int(num)
+
+
+@cuda.jit(device=True)
+def atomic_cast_none(num):
+    return num
+
+
+@cuda.jit(device=True)
+def atomic_binary_1dim_shared(ary, idx, op2, ary_dtype, ary_nelements, binop_func, cast_func, initializer):
```
Comment (stuartarchibald) [inline numba/cuda/tests/cudapy/test_atomics.py:39]: long line, for new code we try and keep to <80 chars https://numba.readthedocs.io/en/stable/developer/contributing.html#coding-conventions

File: numba/cuda/tests/cudapy/test_atomics.py [hunk 3]
```diff
@@ -20,229 +20,154 @@ def skip_unless_cc_32(fn):
...
-    cuda.atomic.add(sm, bin, 1)
+    bin = cast_func (idx[tid] % ary_nelements)
+    binop_func(sm, bin, op2)
     cuda.syncthreads()
     ary[tid] = sm[tid]
 
 
-def atomic_add2(ary):
+@cuda.jit(device=True)
+def atomic_binary_2dim_shared(ary, op2, ary_dtype, ary_shape, binop_func, y_cast_func):
```
Comment (stuartarchibald) [inline numba/cuda/tests/cudapy/test_atomics.py:51]: long line

File: numba/cuda/tests/cudapy/test_atomics.py [hunk 4]
```diff
@@ -20,229 +20,154 @@ def skip_unless_cc_32(fn):
...
+def atomic_add2(ary):
+    atomic_binary_2dim_shared(ary, 1, uint32, (4,8), cuda.atomic.add, atomic_cast_none)
+
+
+def atomic_add3(ary):
+    atomic_binary_2dim_shared(ary, 1, uint32, (4,8), cuda.atomic.add, atomic_cast_to_uint64)
+
+
+def atomic_add_float(ary):
+    atomic_binary_1dim_shared(ary, ary, 1.0, float32, 32, cuda.atomic.add, atomic_cast_to_int, 0.0)
```
Comment (stuartarchibald) [inline numba/cuda/tests/cudapy/test_atomics.py:89]: long lines

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch, and thanks for reviewing @gmarkall . Code changes look good technically, there's just a few stylistic things to take a look at, namely an unused import and a load of long lines (I've not marked them all, but if it's possible in new code to wrap to 80 chars unless absolutely necessary (e.g. html hyperlink) https://numba.readthedocs.io/en/stable/developer/contributing.html#coding-conventions, that'd be great, thanks!).

**Review by gmarkall — APPROVED** [review]
Looks good to me!

**Comment by gmarkall** [comment]
@stuartarchibald Are you happy with moving this to RTM?

**Review by stuartarchibald — APPROVED** [review]
Thanks for the fixes.

**Pull request merged**